File: Python/146 LRU缓存机制.py
"""
146. LRU缓存机制
运用你所掌握的数据结构，设计和实现一个  LRU (最近最少使用) 缓存机制。它应该支持以下操作： 获取数据 get 和 写入数据 put 。

获取数据 get(key) - 如果密钥 (key) 存在于缓存中，则获取密钥的值（总是正数），否则返回 -1。
写入数据 put(key, value) - 如果密钥已经存在，则变更其数据值；如果密钥不存在，则插入该组「密钥/数据值」。当缓存容量达到上限时，它应该在写入新数据之前删除最久未使用的数据值，从而为新的数据值留出空间。

 

进阶:

你是否可以在 O(1) 时间复杂度内完成这两种操作？

 

示例:

LRUCache cache = new LRUCache( 2 /* 缓存容量 */ );

cache.put(1, 1);
cache.put(2, 2);
cache.get(1);       // 返回  1
cache.put(3, 3);    // 该操作会使得密钥 2 作废
cache.get(2);       // 返回 -1 (未找到)
cache.put(4, 4);    // 该操作会使得密钥 1 作废
cache.get(1);       // 返回 -1 (未找到)
cache.get(3);       // 返回  3
cache.get(4);       // 返回  4
"""

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

# ??? 超时了？？？

class LRUCache_timetick:
    def __init__(self, capacity: int):
        self.datas = {}
        #操作计数 用于管理时间
        self.tick = 0
        self.maxSize = capacity

    def get(self, key: int) -> int:
        self.tick += 1
        if key not in self.datas:
            return -1
        else:
            self.datas[key][1] = self.tick
            return self.datas[key][0]

    def put(self, key: int, value: int) -> None:
        self.tick += 1
        if key in self.datas:
            self.datas[key][1] = self.tick
            self.datas[key][0] = value
        else:
            if len(self.datas) >= self.maxSize:
                datalist = list(self.datas.items())
                self.datas.pop(min(datalist,key= lambda x:x[1][1])[0])
            self.datas[key] = [value,self.tick]

class LRUCache:
    def __init__(self, capacity: int):
        self.datas = {}
        #用于维护时间的队列
        self.actionQueue = []
        
        self.maxSize = capacity

    def get(self, key: int) -> int:
        if key not in self.datas:
            return -1
        else:
            self.actionQueue.remove(key)
            self.actionQueue.append(key)
            return self.datas[key]

    def put(self, key: int, value: int) -> None:
        if key in self.datas:
            self.actionQueue.remove(key)
            self.actionQueue.append(key)
            self.datas[key] = value
        else:
            if len(self.datas) >= self.maxSize:
                oldkey = self.actionQueue.pop(0)
                self.datas.pop(oldkey)
            self.actionQueue.append(key)
            self.datas[key] = value

LUR = LRUCache(2)
print(LUR.put(1,1))
print(LUR.put(2,2))
print(LUR.get(1))
print(LUR.put(3,3))
print(LUR.get(2))
print(LUR.put(4,4))
print(LUR.get(1))
print(LUR.get(3))
print(LUR.get(4))

File: CPP/146 LRU缓存机制.cpp
#include "include.hpp"
using namespace std;
/*
题目描述
设计LRU缓存结构，该结构在构造时确定大小，假设大小为K，并有如下两个功能
set(key, value)：将记录(key, value)插入该结构
get(key)：返回key对应的value值
[要求]
set和get方法的时间复杂度为O(1)
某个key的set或get操作一旦发生，认为这个key的记录成了最常使用的。
当缓存的大小超过K时，移除最不经常使用的记录，即set或get最久远的。
若opt=1，接下来两个整数x, y，表示set(x, y)
若opt=2，接下来一个整数x，表示get(x)，若x未出现过或已被移除，则返回-1
对于每个操作2，输出一个答案
示例1
输入
复制
[[1,1,1],[1,2,2],[1,3,2],[2,1],[1,4,4],[2,2]],3
输出
复制
[1,-1]
说明
第一次操作后：最常使用的记录为("1", 1)
第二次操作后：最常使用的记录为("2", 2)，("1", 1)变为最不常用的
第三次操作后：最常使用的记录为("3", 2)，("1", 1)还是最不常用的
第四次操作后：最常用的记录为("1", 1)，("2", 2)变为最不常用的
第五次操作后：大小超过了3，所以移除此时最不常使用的记录("2", 2)，加入记录("4", 4)，并且为最常使用的记录，然后("3", 2)变为最不常使用的记录
备注:
1 \leq K \leq N \leq 10^51≤K≤N≤10 
5
 
-2 \times 10^9 \leq x,y \leq 2 \times 10^9−2×10 
9
 ≤x,y≤2×10 
9
*/

#include <unordered_map>

class LRUCache {
public:
    LRUCache(int capacity) {
        size = capacity;
    }
 
    int count = 0;
    int size = 0;
    int step = 1;
    unordered_map<int,int> data;
    unordered_map<int,int> time;
    
    int get(int key)
    {
        step++;

        if(time[key] != 0)
        {
            time[key] = step;
            return data[key];
        }
        else
        {
            return -1;
        }
    };
    
    void put(int key,int val)
    {
        step++;
        if(time[key] == 0)
            count += 1;
        data[key] = val;
        time[key] = step;
        if(count > size)
        {
            int rmKey = 0;
            int minTime = 99999999;
            for(auto p:time)
            {
                if(p.second != 0 && p.second < minTime)
                {
                    minTime = p.second;
                    rmKey = p.first;
                }
            }
            time[rmKey] = 0;
            count--;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
int main()
{
    Solution s;
    system("pause");
};